### PR TITLE
Lower CPU usage of gulp:watch

### PR DESCRIPTION
**Pull request by giorgiosironi**
To debug https://github.com/elifesciences/issues/issues/4701 on @discodavey's Mac

### CODE REVIEW

**Comment by discodavey** [comment]
strace: test_ptrace_setoptions_for_all: PTRACE_TRACEME doesn't work: Operation not permitted
strace: test_ptrace_setoptions_for_all: unexpected exit status 1

**Comment by discodavey** [comment]
read(8, "\1\0\0\0\0\0\0\0", 1024)       = 8
epoll_wait(3, {{EPOLLIN, {u32=8, u64=8}}}, 1024, 8) = 1
read(8, "\1\0\0\0\0\0\0\0", 1024)       = 8
epoll_wait(3, {{EPOLLIN, {u32=8, u64=8}}}, 1024, 8) = 1
read(8, "\1\0\0\0\0\0\0\0", 1024)       = 8
epoll_wait(3, {{EPOLLIN, {u32=8, u64=8}}}, 1024, 6) = 1
read(8, "\1\0\0\0\0\0\0\0", 1024)       = 8
epoll_wait(3, {{EPOLLIN, {u32=8, u64=8}}}, 1024, 6) = 1
read(8, "\1\0\0\0\0\0\0\0", 1024)       = 8
epoll_wait(3, {{EPOLLIN, {u32=8, u64=8}}}, 1024, 5) = 1
read(8, "\1\0\0\0\0\0\0\0", 1024)       = 8
epoll_wait(3, {{EPOLLIN, {u32=8, u64=8}}}, 1024, 4) = 1
read(8, "\1\0\0\0\0\0\0\0", 1024)       = 8
epoll_wait(3, {{EPOLLIN, {u32=8, u64=8}}}, 1024, 4) = 1
read(8, "\1\0\0\0\0\0\0\0", 1024)       = 8
epoll_wait(3, {{EPOLLIN, {u32=8, u64=8}}}, 1024, 43) = 1
read(8, "\1\0\0\0\0\0\0\0", 1024)       = 8
epoll_wait(3, {{EPOLLIN, {u32=8, u64=8}}}, 1024, 43) = 1
read(8, "\1\0\0\0\0\0\0\0", 1024)       = 8
epoll_wait(3, {{EPOLLIN, {u32=8, u64=8}}}, 1024, 42) = 1
read(8, "\1\0\0\0\0\0\0\0", 1024)       = 8
epoll_wait(3, {{EPOLLIN, {u32=8, u64=8}}}, 1024, 42) = 1
read(8, "\1\0\0\0\0\0\0\0", 1024)       = 8
epoll_wait(3, {{EPOLLIN, {u32=8, u64=8}}}, 1024, 41) = 1
read(8, "\1\0\0\0\0\0\0\0", 1024)       = 8
epoll_wait(3, {{EPOLLIN, {u32=8, u64=8}}}, 1024, 39) = 1
read(8, "\1\0\0\0\0\0\0\0", 1024)       = 8
epoll_wait(3, {{EPOLLIN, {u32=8, u64=8}}}, 1024, 39) = 1
read(8, "\1\0\0\0\0\0\0\0", 1024)       = 8
epoll_wait(3, {{EPOLLIN, {u32=8, u64=8}}}, 1024, 39) = 1
read(8, "\1\0\0\0\0\0\0\0", 1024)       = 8
epoll_wait(3, {{EPOLLIN, {u32=8, u64=8}}}, 1024, 37) = 1
read(8, "\1\0\0\0\0\0\0\0", 1024)       = 8
epoll_wait(3, {{EPOLLIN, {u32=8, u64=8}}}, 1024, 36) = 1
read(8, "\1\0\0\0\0\0\0\0", 1024)       = 8
epoll_wait(3, {{EPOLLIN, {u32=8, u64=8}}}, 1024, 35) = 1
read(8, "\1\0\0\0\0\0\0\0", 1024)       = 8
epoll_wait(3, {{EPOLLIN, {u32=8, u64=8}}}, 1024, 35) = 1
read(8, "\1\0\0\0\0\0\0\0", 1024)       = 8
epoll_wait(3, {{EPOLLIN, {u32=8, u64=8}}}, 1024, 34) = 1
read(8, "\1\0\0\0\0\0\0\0", 1024)       = 8
epoll_wait(3, {{EPOLLIN, {u32=8, u64=8}}}, 1024, 32) = 1
read(8, "\1\0\0\0\0\0\0\0", 1024)       = 8

**Comment by giorgiosironi** [comment]
Some analysis: `gulp.watch` (not `gulp-watch`) uses `vynil-fs` which uses `glob-watcher` which uses `gaze`...

**Comment by giorgiosironi** [comment]
`gaze` passes an `interval` option to [`fs.watchFile()`](https://nodejs.org/api/all.html#fs_fs_watchfile_filename_options_listener) which seems related to polling.

**Comment by giorgiosironi** [comment]
`fs.watchFile` uses a [`StatWatcher` object](https://github.com/nodejs/node/blob/v6.14.3/lib/fs.js). [Rumour](https://stackoverflow.com/a/4482806/91590) has it that it doesn't work well for hundreds of files.

**Comment by giorgiosironi** [comment]
@discodavey ready for a new run.

**Comment by discodavey** [comment]
Updated the branch from last night and this is the results

myhost:pattern-library daveprice$ bin/watch
Starting assets-builder Docker container...
Creating network "pattern-library_default" with the default driver
Creating volume "pattern-library_node_modules" with default driver
Creating volume "pattern-library_source_assets" with default driver
Creating volume "pattern-library_test_build" with default driver
Creating volume "pattern-library_public" with default driver
Creating pattern-library_npm_1 ... done
pattern-library-gulp-watch
Starting ui Docker container...
Starting pattern-library_npm_1      ... done
Creating pattern-library_composer_1       ... done
Creating pattern-library_assets-builder_1 ... done
Creating pattern-library_assets_1         ... done
configuring pattern lab...
your site has been generated...
site generation took 7.3931951522827 seconds and used 20MB of memory...
Starting pattern-library_composer_1 ... done
Starting pattern-library_npm_1      ... done
Starting pattern-library_assets-builder_1 ... done
Starting pattern-library_assets_1         ... done
pattern-library-php-watch
Starting pattern-library_npm_1      ... done
Starting pattern-library_composer_1 ... done
Starting pattern-library_assets-builder_1 ... done
Starting pattern-library_assets_1         ... done
Creating pattern-library_ui-builder_1     ... done
Creating pattern-library_ui_1             ... done
configuring pattern lab...
[09:43:42] Using gulpfile /srv/pattern-library/gulpfile.js
[09:43:42] Starting 'sass:watch'...
[09:43:45] Finished 'sass:watch' after 2.91 s
[09:43:45] Starting 'img:watch'...
[09:43:56] Finished 'img:watch' after 10 s
[09:43:56] Starting 'js:watch'...
[09:43:58] Finished 'js:watch' after 2.17 s
[09:43:58] Starting 'fonts:watch'...
[09:43:58] Finished 'fonts:watch' after 67 ms
[09:43:58] Starting 'watch'...
[09:43:58] Finished 'watch' after 210 μs
your site has been generated...
site generation took 7.6427848339081 seconds and used 20MB of memory...
watching your site for changes...

**Comment by giorgiosironi** [comment]
@davidcmoulton can you try `bin/watch` usage on this branch too, to see any adverse effects? It can take 1s more to notice changes.

File: Dockerfile.assets-builder [hunk 1]
```diff
@@ -27,3 +27,5 @@ RUN node_modules/.bin/gulp --environment ${environment} js
 
 COPY assets/sass/ assets/sass/
 RUN node_modules/.bin/gulp --environment ${environment} generateCss
+
+RUN apt-get update && apt-get install strace
```
Comment (giorgiosironi) [inline Dockerfile.assets-builder]: This will be removed once the investigation is complete.

**Comment by giorgiosironi** [comment]
Current result on @discodavey: load is low enough that it takes 10s rather than minutes for the rebuilding cycle to complete. We can try to get a lower time still.

**Comment by discodavey** [comment]
This has improved things a lot. The average time has now gone down for a JavaScript change from 90-120 secs to 10secs approx and CSS changes are now down to 4-5secs approx from around 20-30secs.

Happy to work with this but if it could be reduced further when time allows that would be great.

**Comment by giorgiosironi** [comment]
I take that as an approval :+1:

**Pull request merged**